src/index.ts

```diff
@@ -74,12 +101,25 @@ async function retryFetch(
 
   try {
     const response = await promiseRetry(async (throwRetryError, retryCount) => {
+      jsonFetchOptions.onRequestStart?.({url: requestUrl, retryCount, ...jsonFetchOptions});
       try {
         const res = await fetch(requestUrl, requestOptions);
         if (shouldRetry(res)) throwRetryError(null);
+        jsonFetchOptions.onRequestEnd?.({
+          responseOrError: res,
+          url: requestUrl,
+          retryCount,
+          ...jsonFetchOptions,
+        });
         return res;
       } catch (err) {
         err.retryCount = retryCount - 1;
+        jsonFetchOptions.onRequestEnd?.({
+          responseOrError: err,
+          url: requestUrl,
+          retryCount,
+          ...jsonFetchOptions,
+        });
```

Comment:
I believe this works, but I find it pretty hard to reason about. You have to make some leaky assumptions about `throwRetryError` and think through multiple control flows. I think it would be simpler and more expressive to simply call `.onRequestEnd()` a single time immediately after line 106 (`const res = ...`). The goal here is to call this hook immediately after making any request.

... Hrm, but this isn't possible, because we intend to pass the `errOrResponse` and we only have access to all possible errors in the `catch` block. Even if we simplified this to only pass `response` even in 'error' cases, this still wouldn't be truly possible because there's a flow where `const res = await fetch(...)` doesn't even resolve, it throws, so we would still need to handle this in the catch... 🤔

Alright, I think this is OK then.

Reply:
Yep, I proved all your thought from second block while coding, I wasn't able to find a less verbose nor cleaner way to do the calls, so so far seems to be the best way 👍🏼